crates/table/src/table.rs

```diff
@@ -530,26 +534,53 @@ impl Table {
         Ok(())
     }
 
+    /// Finds the [`RowPointer`] to the row in `target_table` equal, if any,
+    /// to the row `needle_ptr` in `needle_table`,
+    /// by any unique index in `target_table`.
+    ///
+    /// # Safety
+    ///
+    /// `needle_table.is_row_present(needle_ptr)` must hold.
+    unsafe fn find_same_row_via_unique_index(
+        target_table: &Table,
+        needle_table: &Table,
+        needle_bs: &dyn BlobStore,
+        needle_ptr: RowPointer,
+    ) -> Option<RowPointer> {
+        // Find the smallest unique index.
+        let (cols, idx) = target_table
+            .indexes
+            .iter()
+            .find(|(_, idx)| idx.is_unique())
+            .expect("there is at least one unique index");
+        // Project the needle row to the columns of the index, and then seek.
+        // As this is a unique index, there are 0-1 rows for this key.
+        let needle_row = unsafe { needle_table.get_row_ref_unchecked(needle_bs, needle_ptr) };
+        let key = needle_row.project(cols).expect("needle row should be valid");
+        idx.seek(&key).next()
+    }
+
     /// Insert row identified by `ptr` into the pointer map.
     /// This checks for set semantic violations.
     /// Deletes the row and returns an error if there were any violations.
-    /// Returns the row hash computed.
+    /// Returns the row hash computed, if any.
```

Comment:
Please update this doc comment to describe that this method should be called even if `self.pointer_map.is_none()`. I suggest:

```rust
/// If this table has a [`PointerMap`], insert the row identified by `ptr` into that map.
///
/// This checks for set semantic violations.
/// If a set semantic conflict (i.e. duplicate row) is detected by the pointer map,
/// the row will be deleted and an error returned.
/// If the pointer map confirms that the row was unique, returns the `RowHash` of that row.
///
/// If this table has no [`PointerMap`], returns `Ok(None)`.
/// In that case, the row's uniqueness will be verified by [`Self::insert_into_indices`],
/// as this table has at least one unique index.